iroh/src/node.rs

```diff
@@ -710,33 +749,571 @@ impl<D: BaoMap + BaoReadonlyDb, C: CollectionParser> RpcHandler<D, C> {
+        // TODO: the data we get is validated at this point, but we need to check
+        // that it actually contains the requested ranges. Or DO WE?
+        use iroh_io::AsyncSliceWriter;
+
+        let hash = header.hash();
+        // read the size
+        let (content, size) = header.next().await?;
+        // open the data file in any case
+        let df = entry.data_writer().await?;
+        let mut of = if needs_outboard(size) {
+            Some(entry.outboard_mut().await?)
+        } else {
+            None
+        };
+        // allocate a new id for progress reports for this transfer
+        let id = sender.new_id();
+        sender.send(ShareProgress::Found { id, hash, size }).await?;
+        let sender2 = sender.clone();
+        let on_write = move |offset: u64, _length: usize| {
+            // if try send fails it means that the receiver has been dropped.
+            // in that case we want to abort the write_all_with_outboard.
+            sender2
+                .try_send(ShareProgress::Progress { id, offset })
+                .map_err(|e| {
+                    tracing::info!("aborting download of {}", hash);
+                    e
+                })?;
+            Ok(())
+        };
+        let mut pw = ProgressSliceWriter2::new(df, on_write);
+        // use the convenience method to write all to the two vfs objects
+        let end = content
+            .write_all_with_outboard(of.as_mut(), &mut pw)
+            .await?;
+        // sync the data file
+        pw.sync().await?;
+        // sync the outboard file
+        if let Some(mut of) = of {
+            of.sync().await?;
+        }
+        // actually store the data. it is up to the db to decide if it wants to
+        // rename the files or not.
+        db.insert_complete(entry).await?;
+        // notify that we are done
+        sender.send(ShareProgress::Done { id }).await?;
+        Ok(end)
+    }
+
+    async fn get_missing_ranges_blob(
+        entry: &D::PartialEntry,
+    ) -> anyhow::Result<RangeSet2<ChunkNum>> {
+        use tracing::trace as log;
+        // compute the valid range from just looking at the data file
+        let mut data_reader = entry.data_reader().await?;
+        let data_size = data_reader.len().await?;
+        let valid_from_data = RangeSet2::from(..ByteNum(data_size).full_chunks());
+        // compute the valid range from just looking at the outboard file
+        let mut outboard = entry.outboard().await?;
+        let valid_from_outboard = bao_tree::io::fsm::valid_ranges(&mut outboard).await?;
+        let valid: RangeSet2<ChunkNum> = valid_from_data.intersection(&valid_from_outboard);
+        let total_valid: u64 = valid
+            .iter()
+            .map(|x| match x {
+                RangeSetRange::Range(x) => x.end.to_bytes().0 - x.start.to_bytes().0,
+                RangeSetRange::RangeFrom(_) => 0,
+            })
+            .sum();
+        log!("valid_from_data: {:?}", valid_from_data);
```

Comment:
what does `log!` do?

Reply:
It's just so I can adjust the log level for the entire fn in a single line. `use tracing::trace as log;`